src/coordsim/simulation/simulatorparams.py

```diff
@@ -106,13 +126,33 @@ def update_single_inter_arr_mean(self, new_mean):
     def update_single_predicted_inter_arr_mean(self, new_mean):
         self.predicted_inter_arr_mean = {node_id: new_mean for node_id in self.network.nodes}
 
-    # string representation for logging
-    def __str__(self):
-        params_str = "Simulator parameters: \n"
-        params_str += "inter_arr_mean: {}\n".format(self.inter_arr_mean)
-        params_str += f"deterministic_arrival: {self.deterministic_arrival}\n"
-        params_str += "flow_dr_mean: {}\n".format(self.flow_dr_mean)
-        params_str += "flow_dr_stdv: {}\n".format(self.flow_dr_stdev)
-        params_str += "flow_size_shape: {}\n".format(self.flow_size_shape)
-        params_str += f"deterministic_size: {self.deterministic_size}\n"
-        return params_str
+    def generate_flow_lists(self):
+        """Generate and set dicts of lists of flow arrival, size, dr for the run duration"""
+        if self.deterministic_arrival or self.use_states or self.use_trace:
+            # TODO: implement for deterministic traffic and MMPP?
+            raise NotImplementedError()
+
+        # init empty list for each ingress node: node_id --> []
+        flow_arrival = {ing[0]: [] for ing in self.ing_nodes}
+        # TODO: init flow size & dr
+        # generate flow inter-arrival times for each ingress
+        for ing in flow_arrival.keys():
+            # generate flows for time frame of num_steps
+            while sum(flow_arrival[ing]) < self.run_duration:
```

Comment:
This will generate a list with a sum > run_duration most of the time. It will cause inconsistencies in predicting traffic for the following scenario:

```
flow_arrival_list['pop0'] = [..., 19.5] # Sum = 109.5
# The simulator will wait for 9 time steps into the next run duration
# new flow_arrival_list is generated
flow_arrival_list['pop0'] = [5.2, ...]
# Simulator starts accessing this list after the 9.5 steps from last run finished, possibly causing some elements in the array to be completely ignored. 
```
This is what is causing my index exception because I check whether we reached the last element in the array, which is not always the case

Reply:
True. I think, I could come up with a calculation of the predicted traffic given that list.
But another, more severe, problem is the next run for which we already have flows scheduled; yet we reset the flow list at the beginning of the run. This will make it impossible to keep track of what's happening.

My original idea was to generate the flows for the entire episode up front - not just for one run. That way we wouldn't have problems with flows overlapping in different runs.

But we don't know the episode length (or the number of runs) in the simulator, do we? Could we just pass it?

Maybe more elegant to just generate more per run, but not throwing old data away. And we'll need to generate 2x run_duration to be able to use 1x run_duration for the current run ahead and 1x run_duration for predicting the flows of the next run.
I'll have another try.

Reply:
The generation for the entire episode would be possible in a straightforward way only for a fixed mean. It would not work for MMPP and trace files since we wouldn't know the upcoming state or mean changes. 

Since this is an 'oracle' prediction, let's try to make it 100% correct if possible. Did you have a look at my implementation? I think it would be better if we can focus our ideas on fixing one of the implementations to prevent redundant work from both of us. Let me know if I should close my PR and focus on improving this PR with you instead.